azure/services/virtualmachines/client.go

```diff
@@ -62,38 +71,150 @@ func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, vmName string
 	return ac.virtualmachines.Get(ctx, resourceGroupName, vmName, "")
 }
 
-// CreateOrUpdate the operation to create or update a virtual machine.
-func (ac *AzureClient) CreateOrUpdate(ctx context.Context, resourceGroupName, vmName string, vm compute.VirtualMachine) error {
+// CreateOrUpdateAsync creates or updates a virtual machine asynchronously.
+// It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter) (interface{}, azureautorest.FutureAPI, error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualmachines.AzureClient.CreateOrUpdate")
 	defer done()
 
-	future, err := ac.virtualmachines.CreateOrUpdate(ctx, resourceGroupName, vmName, vm)
+	vm, err := ac.vmParams(ctx, spec)
 	if err != nil {
-		return err
+		return nil, nil, errors.Wrapf(err, "failed to get desired parameters for virtual machine %s", spec.ResourceName())
+	} else if vm == nil {
+		// nothing to do here.
```

Comment:
self-review: maybe this should return the result as the existing vm instead of returning nil so we can act upon the existing VM even if it already exists

Reply:
Could be added in the future if needed.